Climate 1.0

pvizeli opened the pull request:
## Breaking Change:

* All operation are now defined by the core
* Abstract operation mode and current action for devices they support this (only for viewing on UI)
* Hold modes called now presets
* Away mode move into hold modes because it is what it is and manufactory use it like that
* Remove on/off in flavor of hvac state

**Platforms:**

Most platform will support now a lot more features. But Somethings was also cleaned out:

* OEM - Remove fake away
* Hive - Hot water is not a climate device @KJonline @Rendili (use the new component for hot water) / boost use now preset instead hack into auxiliary
* Radiotherm - remove fake away
* Nuheat - service renamed from `climate.nuheat_resume_program` to `nuheat.resume_program`
* generic-thermostat - rename `initial_operation_mode` to `initial_hvac_mode`

## Description:

Looks like a small change but they are massive. I look over all climate API and also external Interfaces like Voice Assistant, and that is the result they should be able to cover our cases.

Most of the problem was that a lot of stuff was mixed because of history for that component. Like Away mode, they were initially added and later as the hold modes come too, they were never strict merged back. Some integration was added hacks to offload it from hold mode.

That is the result of hours and hours research of what we have and what exists and how the device manufacture implement things.

I don't think our old climate was bad because no platform supports so many different devices and I learned a lot from that code. The problem was only, that no one had the overview to make a clean setup.

Implement the subset of over 100 comments from https://github.com/home-assistant/architecture/issues/22

This pull request includes an updated UI to work with the new schema.

## AWAY

Some HA hardware interfaces (components) use a fake away. That is something that is not part of an interface schema. Logic abstraction or components level 2 allowed to fake logic but not on level 1 (hardware interfaces).

## Upgrade guide

The main thing that is new is presets Things like eco mode, away mode, hold mode, manual mode, etc are now all presets and set via `set_preset_mode` function. 

- New `set_preset_mode`/`async_set_preset_mode` replaces the following (including their async variants):
   - `turn_away_mode_on`
   - `turn_away_mode_off`
   - `set_hold_mode`
   - `turn_on`
   - `turn_off`
- `supported_features`:
    - `SUPPORT_TARGET_TEMPERATURE_LOW` + `SUPPORT_TARGET_TEMPERATURE_HIGH` -> `SUPPORT_TARGET_TEMPERATURE_RANGE`
    - Removed `SUPPORT_TARGET_HUMIDITY_HIGH`, `SUPPORT_TARGET_HUMIDITY_LOW`
    - Removed `SUPPORT_OPERATION_MODE`
    - Removed `SUPPORT_HOLD_MODE`
    - Removed `SUPPORT_SWING_MODE`
    - Removed `SUPPORT_AWAY_MODE`
    - Removed `SUPPORT_ON_OFF` (implement HVAC_MODE_HEAT + HVAC_MODE_OFF instead)
- `is_on` removed  (implement HVAC_MODE_HEAT + HVAC_MODE_OFF instead)
- `STATE_IDLE` -> `HVAC_STATE_OFF`
- State machine attribute `ATTR_OPERATION_MODE` is now the state of the entity
- Properties that are the same:
  - `name`
  - `precision_unit`
  - `temperature_unit`
  - `current_humidity`
  - `target_humidity`
  - `current_temperature`
  - `target_temperature`
  - `target_temperature_step`
  - `target_temperature_high`
  - `target_temperature_low`
  - `min_temp`
  - `max_temp`
  - `min_humidity`
  - `max_humidity`
- `current_operation` -> `hvac_mode`(limited to one of `HVAC_MODES` const)
- `operation_list` -> `hvac_modes` (subset of `HVAC_MODES` const)
- `hvac_action` **new**. The current action of the device (heating, etc. See `CURRENT_HVAC_*` consts)
- `preset_mode` **new** The current preset mode (home, away, temp) (one of `preset_modes`)
- `preset_modes` **new** The available preset modes (prefer one of `PRESET_*` const, custom values allowed)
- `is_aux_heat_on` -> `is_aux_heat`
- `current_fan_mode` -> `fan_mode`  (one of `fan_modes`)
- `fan_list` -> `fan_modes` (prefer using `FAN_*` constants from `climate/const.py`, custom values allowed)
- `current_swing_mode` -> `swing_mode` (one of `swing_modes`)
- `swing_list` -> `swing_modes` (prefer `SWING_` constants, custom values allowed)
- `set_operation_mode` -> `set_hvac_mode`

### Migration

* [x] homematic
* [x] demo
* [x] xs1
* [x] hive
* [x] mill
* [x] lcn
* [x] oem
* [x] toon
* [x] deconz
* [x] tesla
* [x] vera
* [x] zwave
* [x] velbus
* [x] wink
* [x] dyson
* [x] tefiac
* [x] tado
* [x] mqtt
* [x] venstar
* [x] zhong_hong
* [x] tuya
* [x] spider
* [x] radiotherm
* [x] sensibo
* [x] smartthings
* [x] melissa
* [x] fibaro
* [x] mysensors
* [x] nest
* [x] ecobee
* [x] elkm1
* [x] flexit
* [x] coolmaster
* [x] daikin
* [x] honeywell #24257
* [x] proliphix
* [x] ephember
* [x] incomfort
* [x] homekit
* [x] homekit_controller
* [x] geniushub  #24191
* [x] touchline
* [x] stiebel_eltron
* [x] opentherm_gw
* [x] nuheat
* [x] netatmo
* [x] modbus
* [x] maxcube
* [x] knx
* [x] homematicip_cloud
* [x] heatmiser
* [x] generic_thermostat
* [x] fritzbox
* [x] evohome #24491
* [x] esphome
* [x] eq3btsmart
* [x] ambiclimate #24911

### Integration

* [x] Alexa
* [x] Google Assistant

pvizeli committed:
Climate 1.0 / part 1/2/3

balloob commented on homeassistant/components/climate/__init__.py:
```diff
@@ -290,15 +247,20 @@ def target_humidity(self):
 
+    @property
+    def current_operation(self):
+        """Return the current action they is running."""
```
```suggestion
        """Return the current action that is running."""
```

balloob commented on homeassistant/components/climate/const.py:
```diff
@@ -1,19 +1,75 @@
+# other related mechanism. User is not able to adjust the temperature
+OPERATION_MODE_AUTO = 'auto'
+
+# Device is in Dry/Huminity mode
```
```suggestion
# Device is in Dry/Humidity mode
```

balloob commented on homeassistant/components/climate/const.py:
```diff
@@ -1,19 +1,75 @@
+
+
+# This are support current states of HVAC
+OPERATION_CURRENT_OFF = 'off'
```
Let's name them after the property, so `CURRENT_OPERATION_OFF`

elupus replied:
wait. the property is "operation_mode".  current_operation is a free value to represent what is currently going on in the device. ie not limited by this list?

elupus replied:
Nevermind. See that they are called OPERATION_CURRENT. So indeed what is going one. Scratch previous comment

balloob commented on homeassistant/components/climate/__init__.py:
```diff
@@ -187,13 +150,7 @@ class ClimateDevice(Entity):
-        if self.is_on:
-            return STATE_ON
-        return None
+        return self.hvac_state
```
Should this be

```suggestion
        return self.operation_mode
```

?

pvizeli replied:
Is fixed but I forget to push it :D

balloob commented on homeassistant/components/climate/const.py:
```diff
@@ -1,19 +1,75 @@
+
+
+# Device is running an energy-saving mode
+HOLD_MODE_ECO = 'eco'
```
So hold mode is what in some systems is called a profile?

pvizeli replied:
Yes and no. Profiles are only presetting of temperatures, but hold modes have their own life/circle or actions.

pvizeli replied:
But I'm fine to rename as profile

balloob replied:
I'm not proposing to rename it, just wanted to clarify.

balloob commented on homeassistant/components/climate/const.py:
```diff
@@ -1,19 +1,75 @@
+
+# This are support current states of HVAC
+OPERATION_CURRENT_OFF = 'off'
+OPERATION_CURRENT_HEAT = 'heat'
```
We should make these values present participle: off, heating, cooling, drying, idle, fan_only

balloob commented on homeassistant/components/climate/__init__.py:
```diff
@@ -325,13 +287,13 @@ def target_temperature_low(self):
         return None
```
I wonder if we should collapse the temperatures? Have the entity return either a single value or a list? Now theoretically it could return `None`

pvizeli replied:
It's how it works with a supported feature. But I agree, most of the entity are focus to use it on UI instead what is the best way to handle it internally

pvizeli replied:
What do you prefer? I think for influxdb and other things like that is a flat entity not bad like we need on UI

elupus commented on homeassistant/components/climate/__init__.py:
```diff
@@ -290,15 +247,20 @@ def target_humidity(self):
-    def current_operation(self):
-        """Return current operation ie. heat, cool, idle."""
+    def operation_mode(self):
+        """Return current operation ie. heat, cool."""
```
"""Return operation mode ie. heat, cool.""" lets stick to "operation mode" and "current operation" as the two discrete data points.

elupus commented on homeassistant/components/climate/__init__.py:
```diff
@@ -325,13 +287,13 @@ def target_temperature_low(self):
     @property
-    def is_away_mode_on(self):
-        """Return true if away mode is on."""
+    def current_hold_mode(self):
```
current? if current is the running state for operation, same should apply here. ie it should just be "hold_mode"

elupus commented on homeassistant/components/climate/const.py:
```diff
@@ -1,19 +1,75 @@
 ATTR_MIN_TEMP = 'min_temp'
 ATTR_OPERATION_LIST = 'operation_list'
 ATTR_OPERATION_MODE = 'operation_mode'
+ATTR_OPERATION_CURRENT = 'operation_current'
```
should this then not match "current_operation" as attribute?

MartinHjelmare commented on homeassistant/components/climate/__init__.py:
```diff
@@ -96,7 +63,7 @@
         vol.Inclusive(ATTR_TARGET_TEMP_LOW, 'temperature'): vol.Coerce(float),
         vol.Optional(ATTR_ENTITY_ID): cv.comp_entity_ids,
-        vol.Optional(ATTR_OPERATION_MODE): cv.string,
+        vol.Optional(ATTR_OPERATION_MODE): vol.In(OPERATION_MODES),
```
I think we should remove setting operation from the temperature service.

The reason we added it was to allow scenes to set both temperature and operation. We don't need this anymore after the change of reproduce state implementation, where each component defines how the state attributes should be reproduced using services.

pvizeli replied:
Well, I agree, but maybe you want cool and set a new temperature in one call?

balloob replied:
I think that we should either collapse all services in 1, or we don't .

MartinHjelmare commented on homeassistant/components/climate/const.py:
```diff
@@ -1,19 +1,75 @@
+OPERATION_MODE_COOL = 'cool'
+
+# The device supports heating/cooling to a range
+OPERATION_MODE_HEATCOOL = 'heatcool'
```
```py
OPERATION_MODE_HEAT_COOL = 'heat_cool'
```

MartinHjelmare commented on homeassistant/components/climate/const.py:
```diff
@@ -1,19 +1,75 @@
+OPERATION_CURRENT_COOL = 'cool'
+OPERATION_CURRENT_DRY = 'dry'
+OPERATION_CURRENT_IDLE = 'idle'
+OPERATION_CURRENT_FAN_ONLY = 'fan_only'
```
I think we shouldn't call this the same as the operation mode which is also called `'fan_only'`.

pvizeli replied:
`fan`

pvizeli commented:
I'm currently not happy about the hold stuff. Will overthink that part again

MartinHjelmare commented:
I like preset.

elupus commented on homeassistant/components/climate/__init__.py:
```diff
@@ -290,13 +247,18 @@ def target_humidity(self):
-        """Return current operation ie. heat, cool, idle."""
+    def hvac_state(self):
+        """Return current hvac operation ie. heat, cool mode"""
+        return None
```
Not sure you where quite finished here. But now it's somewhat confusing. We have hvac_[state] hvac_[modes] and [current_operation] quite unclear how they relate. Also comments make use of "current" clause.

pvizeli replied:
Yeah, need to fix the comment and update later the developer page. current_operation has no effect on everything. That was a big wish on architecture that you can see the current action of device on UI.

pvizeli replied:
That is also the reason why I don't update it to hvac to avoid that. Also, only 10% of devices support that feature.

elupus replied:
That only 10% support it now doesn't mean only 10% can support it. It was not an allowed behaviour before.

presslab-us commented:
Zigbee thermostats support a true "hold" mode which means that it does not execute the internal schedule, it just holds the current setpoints.  It also optionally allows this hold mode to self disable after some configurable time.

I think it would be nice to at least support this hold mode somehow.  Could the "preset" accommodate this?

pvizeli commented:
@presslab-us the old climate try to integrate every climate native. Now we have our own schema and interfaces can wrap that to the device. I think in future `presets` should support a time range because some devices can use that.

elupus commented:
I just want to suggest again that hvac_state is renamed to hvac_mode. You have hvac_mode(s) plurar for the list, so the current aught to be the singular.

I'd suggest the same for presets and preset.  Or preset_mode and preset_modes for list. That way it's clear they belong together.

balloob commented on homeassistant/components/climate/__init__.py:
```diff
@@ -96,20 +62,20 @@
+SET_HVAC_MODE_SCHEMA = vol.Schema({
     vol.Optional(ATTR_ENTITY_ID): cv.comp_entity_ids,
-    vol.Required(ATTR_OPERATION_MODE): cv.string,
+    vol.Required(ATTR_HVAC_MODE): vol.In(HVAC_MODES),
```
This is genius. Finally a way to stop people from trying to work around our pre-defined constants.

balloob commented on homeassistant/components/climate/__init__.py:
```diff
@@ -240,308 +187,262 @@ def state_attributes(self):
-            data[ATTR_HOLD_MODE] = self.current_hold_mode
+        if supported_features & SUPPORT_PRESET_MODE:
+            data[ATTR_PRESET_MODE] = self.preset_mode
+            if self.preset_list:
```
```suggestion
            if self.preset_list is not None:
```

balloob replied:
Not sure how list is generated might not want to retrieve property twice

zxdavb commented:
I am trying to migrate **evohome** to the new climate platform, PR #23969

I have a difficulty with
```python
supported_features = SUPPORT_TARGET_TEMPERATURE | SUPPORT_CURRENT_HVAC | SUPPORT_PRESET_MODE
````
In the Web UI (/states), I only get: **Fan mode** , and **Swing mode** .  Understandably, both drop-down lists are empty.

I _do_ get the **Target Temperature** in the UI.

If I have `supported_features = 2**16-1`, then I get the full gamut in the UI, but the **Operation** list is empty.
````

balloob commented on homeassistant/components/climate/__init__.py:
```diff
@@ -239,309 +185,258 @@ def state_attributes(self):
-    def current_operation(self):
-        """Return current operation ie. heat, cool, idle."""
-        return None
+    def hvac_state(self) -> str:
```
@pvizeli I like @elupus suggestion of renaming this to `hvac_mode`.

balloob commented on homeassistant/components/climate/__init__.py:
```diff
@@ -239,309 +185,258 @@ def state_attributes(self):
     @property
-    def operation_list(self):
-        """Return the list of available operation modes."""
+    def current_hvac(self) -> Optional[str]:
```
It's a bit weird to add `current` since all properties represent the current state. Maybe we should call this `hvac_operation` or `hvac_action` ? (like action more I think)

balloob commented on homeassistant/components/climate/__init__.py:
```diff
@@ -239,309 +185,258 @@ def state_attributes(self):
     @property
-    def current_hold_mode(self):
-        """Return the current hold mode, e.g., home, away, temp."""
+    def preset_list(self) -> Optional[List[str]]:
```
Should this be `preset_modes` to be in line with `hvac_modes`

balloob commented on homeassistant/components/climate/__init__.py:
```diff
@@ -239,309 +185,258 @@ def state_attributes(self):
 
     @property
-    def swing_list(self):
+    def swing_list(self) -> Optional[List[str]]:
```
`swing_modes`

balloob commented on homeassistant/components/climate/__init__.py:
```diff
@@ -239,309 +185,258 @@ def state_attributes(self):
 
     @property
-    def fan_list(self):
+    def fan_list(self) -> Optional[List[str]]:
```
`fan_modes`

balloob commented:
Sorry for making comments about renaming properties this late. Let me know when you're done updating integrations and I will take care of the renames myself.

zxdavb commented:
@pvizeli Could comment on my post above? https://github.com/home-assistant/home-assistant/pull/23899#issuecomment-495935769

I am happy to migrate **evohome**, but I am a bit stuck.

pvizeli commented:
@zxdavb the UI don't support the new backend currently

Hedda commented:
@pvizeli @Adminiuga Does the ZHA (Zigbee Home Automation) component need a migration too?

zha => https://www.home-assistant.io/components/zha/

zxdavb commented:
@pvizeli the EU half of the **honeywell** component is to be deprecated by 0.96.
https://github.com/home-assistant/home-assistant/blob/fe9b36f65a778f91978f4b45b39895c2f6ec560d/homeassistant/components/honeywell/climate.py#L56-L58

If you're happy for me to do that, I can migrate the US half at the same time.

I can also migrate **incomfort** and **genishub** and well as **evohome**, I just am a bit short on time, is all.

OttoWinter commented on homeassistant/components/climate/__init__.py:
```diff
@@ -239,309 +185,258 @@ def state_attributes(self):
 
     @property
-    def fan_list(self):
+    def fan_modes(self) -> Optional[List[str]]:
```
Should we name all these "available options" properties as just the plural form of the word?

For example: `fan_mode` and `fan_modes`.

I fear this will lead to bugs because the two are very difficult to distinguish when looking over code.

Could we instead name them `_list` so for example `fan_mode_list`

pvizeli replied:
Or `fan_support` ? @balloob

Adminiuga commented:
> Does the ZHA (Zigbee Home Automation) component need a migration too?

@Hedda ATM there's no climate in ZHA component, but the one in the works would need migration, yes.

presslab-us commented:
@Hedda @Adminiuga Are you thinking of the one I am working on?
https://github.com/presslab-us/home-assistant/tree/zha-thermostat
Or is there another one?

Adminiuga commented:
@presslab-us I just wasn't sure if it was you or someone else working on their version based on https://github.com/Adminiuga/home-assistant/tree/ac/new-zha-climate

zxdavb commented:
Is there any appetite to adding more `PRESET`s, like a `PRESET_CUSTOM` or `PRESET_ECO_PLUS`?

zxdavb commented:
@pvizeli Is there any appetite to adding more `PRESET`s, like a 
 - `PRESET_CUSTOM`
 - `PRESET_ECO_PLUS`
 - `PRESET_MANUAL` or `PRESET_OVERRIDE` - zone doing it's own thing rather than according to controller (like a `PRESET_BOOST`, but temp could be set lower rather than higher)
 - `PRESET_OCCUPANCY` - temp according to movement in the room

zxdavb commented:
How do you turn off a `PRESET` mode?

pvizeli commented:
Preset None should be reset/clean a preset

zxdavb commented:
> Preset None should be reset/clean a preset

Would that work, given `preset_mode: str`?
https://github.com/home-assistant/home-assistant/blob/c1eb19321c305eeb24cecbbf3406ff1c30385439/homeassistant/components/climate/__init__.py#L370-L375

When I try calling `climate.set_preset_mode` with:
```json
{
  "entity_id": "climate.room_1_1",
  "preset_mode": null
}
```

I get:
```
Traceback (most recent call last):
  File "/home/dbonnes/home-assistant/homeassistant/components/websocket_api/commands.py", line 121, in handle_call_service
    connection.context(msg))
  File "/home/dbonnes/home-assistant/homeassistant/core.py", line 1121, in async_call
    processed_data = handler.schema(service_data)
  File "/home/dbonnes/home-assistant/venv/lib/python3.6/site-packages/voluptuous/schema_builder.py", line 267, in __call__
    return self._compiled([], data)
  File "/home/dbonnes/home-assistant/venv/lib/python3.6/site-packages/voluptuous/schema_builder.py", line 589, in validate_dict
    return base_validate(path, iteritems(data), out)
  File "/home/dbonnes/home-assistant/venv/lib/python3.6/site-packages/voluptuous/schema_builder.py", line 427, in validate_mapping
    raise er.MultipleInvalid(errors)
voluptuous.error.MultipleInvalid: string value is None for dictionary value @ data['preset_mode']
```

Rendili commented:
Hi @pvizeli ,

I have a few of questions, and a couple of comments on the updated Hive code.

At the moment the hotwater code for Hive has been removed from the Hive climate component, what are your plans for keeping Hive hotwater support after this change has gone live?

Does the water_heater component now support a boost? last time I looked at whether I could move Hive hotwater to the new water_heater component it was missing any boost functionality.

Will the new updates to climate support passing a duration and a temperature as part of calling boost?
At the moment it has been hardcoded for Hive climate to boost for 30 minutes at half a degree higher than current temperature when the aux is invoked.

Rendili commented on homeassistant/components/hive/climate.py:
```diff
@@ -43,21 +45,11 @@ def __init__(self, hivesession, hivedevice):
-
-        self.session.entities.append(self)
+        self.data_updatesource = 'Heating.{}'.format(self.node_id)
+        self._unique_id = '{}-Heating'.format(self.node_id)
```
We need to keep the original format so that it aligns to the other Hive components for when new data is updated from the cloud to the cache from one Hive component, all the other Hive components refresh their data from the latest cached data

pvizeli replied:
sure, but inside `__init__` is a bad design pattern, so I move it out.

Rendili commented on homeassistant/components/hive/climate.py:
```diff
@@ -43,21 +45,11 @@ def __init__(self, hivesession, hivedevice):
         """Initialize the Climate device."""
         self.node_id = hivedevice["Hive_NodeID"]
         self.node_name = hivedevice["Hive_NodeName"]
-        self.device_type = hivedevice["HA_DeviceType"]
```
We need to keep DeviceType to support the original format for the updatesource and unique_id

pvizeli replied:
please read the code, I change nothing on that handling

Rendili replied:
@pvizeli  the line ```self.device_type = hivedevice["HA_DeviceType"] ``` looks to be missing from the new code, this is needed to support the correct `self.data_updatesource` value and format

Rendili commented on homeassistant/components/hive/climate.py:
```diff
@@ -81,26 +73,38 @@ def supported_features(self):
     def handle_update(self, updatesource):
         """Handle the new update request."""
-        if '{}.{}'.format(self.device_type, self.node_id) not in updatesource:
+        if 'Heating.{}'.format(self.node_id) not in updatesource:
```
We need to keep the original format so that it aligns to the other Hive components for when new data is updated from the cloud to the cache from one Hive component, all the other Hive components refresh their data from the latest cached data

pvizeli replied:
like I say, there is no change

Rendili replied:
@pvizeli  unless I am reading it wrong the change is going from:

```if '{}.{}'.format(self.device_type, self.node_id) not in updatesource: ```

to

```if 'Heating.{}'.format(self.node_id) not in updatesource:```


Likewise on the code above the following changes:

```self.data_updatesource = '{}.{}'.format(self.device_type, self.node_id) ```

changes to 

```self.data_updatesource = 'Heating.{}'.format(self.node_id) ```

Which again will not align with the other Hive components code.


So instead of the `self.device_type` in the format it is instead a static word "Heating", this will then not align with the other components for Hive

zxdavb commented:
@pvizeli I have provided 2 migrations, above (honeywell, geniushub). Before I do the other two (evohome, incomfort), could you accept these two?  That way, I will know of I am doing it right.

Also, did you have anything to add re: `PRESET = None`?

pvizeli commented:
@zxdavb that issue can easily be fixed

OnFreund commented:
@pvizeli I'm not sure I got this right (it's a large PR) but if I understand correctly, there's now no way to get the preset mode when the device is off? (since `off` is a state just like `heat`). If that's indeed the case, I think we're missing out critical functionality here.

Adminiuga commented:
@OnFreund what is the use case you have in mind? To me, if the climate is off, there can't be any presets active. As an example, there no light color, if light bulb is off, same with climate

OnFreund commented:
@Adminiuga many A/C systems can be present to a given mode/temperature combo and then simply turned on when needed (BTW `off` as an operating mode feels as arbitrary as `off` as a target temp).
Example use cases:

- An automation that turns the unit on and relies on it being preset to the correct values. Such an automation would now have to figure out the operating mode as well.

And the exact opposite:
- An automation that sets the presets according to some criteria, so that a human can simply turn the device on when they want.

OnFreund commented:
Some more examples:
- Making sure a 2 pipe system has all units set to the same mode
- Making sure (while the device is off) that no one set temps to crazy values, and correcting if needed (if operating mode doesn't make sense when the device is off, so does target temp)

michaeldavie commented:
> @pvizeli I'm not sure I got this right (it's a large PR) but if I understand correctly, there's now no way to get the preset mode when the device is off? (since `off` is a state just like `heat`). If that's indeed the case, I think we're missing out critical functionality here.

I agree. This is the way Sensibo functions, for example. The system is off in the picture below, but you can see the current mode, temperature, and fan speed settings.

![2019-06-15 11 03 16](https://user-images.githubusercontent.com/16008714/59553138-c10c0900-8f5e-11e9-86b7-0b5f80cb9a10.png)

balloob commented:
I don't think it's critical to know that a thermostat would be in heating mode if it wasn't off, because it's off. Likewise, we don't know that a light would turn to X if we turn it on.

OnFreund commented:
@balloob I don't think the comparison to light is the right one, first and foremost because many, if not most, climate systems to save their settings offline, but also because:
* They have way more complicated settings
* The impact of turning on with the wrong settings is more severe (especially if the system controls more than one unit, as in the 2-pipe example)
* Based on what I've seen so far, every smart climate and smart home system I've seen supports this functionality.

This was requested by several people in the arch topic discussing this, and has been brought up by two people in this PR, so it's at least worth a discussion.
All of the 4 use cases I described above are ones that I use, and the proposed PR would make impossible. It will also make working with climate units extra complicated, and I don't see the big gain in dropping support for on/off, and offline settings.

balloob commented:
We're not going to change the architecture anymore. This is it. Once this is merged, we will accept new architecture issues debating how to go from there. But we're not going to do any changes anymore to the current design, nor should it be discussed in this PR. You're welcome to open a new architecture issue.

balloob commented:
Tagging code owners for integrations with climate platforms. Please help migrate your platform to the new Climate 1.0 architecture:

| integration | Code owners |
| -- | -- |
| spider | @peternijssen |
| sensibo | @andrey-git |
| smartthings | @andrewsayre |
| melissa | @kennedyshead |
| nest | @awarecan |
| coolmaster | @OnFreund |
| daikin | @fredrike, @rofrantz |
| ephember | @ttroy50 |
| incomfort | @zxdavb |
| homekit_controller | @Jc2k |
| stiebel_eltron | @fucm |
| knx | @Julius2342 |
| evohome | @zxdavb |
| esphome | @OttoWinter |
| eq3btsmart | @rytilahti |
| ambiclimate | @danielhiversen |

andrewsayre commented:
Sure happy to help—do we have anything that explains what changed/needs to changed to be 1.0 compatible?

shbatm commented:
@balloob -- What is your preference for ISY994 (new platform -- #20970) do you want me to sit on the old PR until this is implemented, or update and send a PR to this branch?

balloob commented:
@shbatm if there is no climate platform yet, let's not add anything new. This is going to be a monster PR without introducing anything new.

balloob commented:
The following integrations have no registered code owners. Looking for any help that we can get. The goal will be to get this all done before 96 beta on July 10.

balloob commented:
Upgrade guide:

- Properties that are the same:
  - `name`
  - `precision_unit`
  - `temperature_unit`
  - `current_humidity`
  - `target_humidity`
  - `current_temperature`
  - `target_temperature`
  - `target_temperature_step`
  - `target_temperature_high`
  - `target_temperature_low`
  - `min_temp`
  - `max_temp`
  - `min_humidity`
  - `max_humidity`
- `current_operation` -> `hvac_mode`(limited to one of `HVAC_MODES` const)
- `operation_list` -> `hvac_modes` (subset of `HVAC_MODES` const)
- `hvac_action` **new**. The current action of the device (heating, etc. See `CURRENT_HVAC_*` consts)
- `preset_mode` **new** The current preset mode (home, away, temp) (limited to one of `PRESET_*` const)
- `preset_modes` **new** The available preset modes (subset of `PRESET_*` const)
- `is_aux_heat_on` -> `is_aux_heat`
- `current_fan_mode` -> `fan_mode`  (limited to one of the `FAN_*` constants from `climate/const.py`)
- `fan_list` -> `fan_modes` (limited to subset of `FAN_*` constants from `climate/const.py`)
- `current_swing_mode` -> `swing_mode` (limited to `SWING_*` constants)
- `swing_list` -> `swing_modes` (subset of `SWING_*` constants)
- `set_operation_mode` -> `set_hvac_mode`
- New `set_preset_mode`/`async_set_preset_mode` replaces the following (including their async variants):
   - `turn_away_mode_on`
   - `turn_away_mode_off`
   - `set_hold_mode`
   - `turn_on`
   - `turn_off`

pvizeli commented:
`preset_modes`, `fan_modes`, `swing_modes` are open @balloob / I looks only that I limited to the subset of constants because 90% of climate use the same/similar names.

`preset_modes`: A bit special, because some things in component const like away/eco should be reused, but it's allowed to extend it if that would make sense.

All that are only secondary lists in UI. The main functionality is the new HVAC handling.

pvizeli committed:
fix flake

balloob commented:
That's fine

balloob committed:
Lint

danielperna84 commented on homeassistant/components/homematic/climate.py:
```diff
@@ -66,40 +60,54 @@ def temperature_unit(self):
+        # Simple devices
+        if self._data.get("BOOST_MODE"):
+            return HVAC_MODE_AUTO
+        return HVAC_MODE_HEAT
```
I don't know how those new modes apply here, but from the name it seems as if the boost mode corresponds to `HVAC_MODE_HEAT`, and `HVAC_MODE_AUTO` is the _regular_ mode.

pvizeli replied:
because `AUTO` means that you can control the temperature at this moment.

balloob replied:
wouldn't that be MODE_HEAT_COOL  ?

balloob committed:
Update Google Assistant

Danielhiversen committed:
ambiclimate to climate 1.0 (#24911)

balloob committed:
Fix Alexa

balloob committed:
Lint

balloob committed:
Migrate zhong_hong

balloob committed:
Migrate tuya

zxdavb committed:
Migrate honeywell to new climate schema (#24257)

* Update one

* Fix model climate v2

* Cleanup p4

* Add comfort hold mode

* Fix old code

* Update homeassistant/components/climate/__init__.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

* Update homeassistant/components/climate/const.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

* First renaming

* Rename operation to hvac for paulus

* Rename hold mode to preset mode

* Cleanup & update comments

* Remove on/off

* Fix supported feature count

* Update services

* Update demo

* Fix tests & use current_hvac

* Update comment

* Fix tests & add typing

* Add more typing

* Update modes

* Fix tests

* Cleanup low/high with range

* Update homematic part 1

* Finish homematic

* Fix lint

* fix hm mapping

* Support simple devices

* convert lcn

* migrate oem

* Fix xs1

* update hive

* update mil

* Update toon

* migrate deconz

* cleanup

* update tesla

* Fix lint

* Fix vera

* Migrate zwave

* Migrate velbus

* Cleanup humity feature

* Cleanup

* Migrate wink

* migrate dyson

* Fix current hvac

* Renaming

* Fix lint

* Migrate tfiac

* migrate tado

* Fix PRESET can be None

* apply PR#23913 from dev

* remove EU component, etc.

* remove EU component, etc.

* ready to test now

* de-linted

* some tweaks

* de-lint

* better handling of edge cases

* delint

* fix set_mode typos

* apply PR#23913 from dev

* remove EU component, etc.

* ready to test now

* de-linted

* some tweaks

* de-lint

* better handling of edge cases

* delint

* fix set_mode typos

* delint, move debug code

* away preset now working

* code tidy-up

* code tidy-up 2

* code tidy-up 3

* address issues #18932, #15063

* address issues #18932, #15063 - 2/2

* refactor MODE_AUTO to MODE_HEAT_COOL and use F not C

* add low/high to set_temp

* add low/high to set_temp 2

* add low/high to set_temp - delint

* run HA scripts

* port changes from PR #24402

* manual rebase

* manual rebase 2

* delint

* minor change

* remove SUPPORT_HVAC_ACTION

balloob committed:
Migrate radiotherm

balloob committed:
Convert touchline

balloob committed:
Migrate flexit

balloob committed:
Migrate nuheat

balloob committed:
Migrate maxcube

balloob committed:
Fix names maxcube const

balloob committed:
Migrate proliphix

balloob committed:
Migrate heatmiser

balloob committed:
Migrate fritzbox

balloob committed:
Migrate opentherm_gw

michaeldavie commented:
`SUPPORT_SWING_MODE` is still at https://github.com/home-assistant/home-assistant/blob/287608c349c03530def4bd0d8a224b32a2835fd9/homeassistant/components/climate/const.py#L123. Should it be?

balloob committed:
Migrate venstar

Swamp-Ig committed:
Migrate daikin

balloob committed:
Migrate modbus

balloob committed:
Fix elif

SukramJ committed:
Migrate Homematic IP Cloud to climate-1.0 (#24913)

* hmip climate fix

* Update hvac_mode and preset_mode

* fix lint

* Fix lint

peternijssen commented:
Will look at my spider component. I can test against the climate-1.0 branch to see if everything functions correctly I suppose.

pvizeli committed:
Migrate generic_thermostat

pvizeli committed:
Merge branch 'climate-1.0' of https://github.com/home-assistant/home-assistant into climate-1.0

zxdavb committed:
Migrate incomfort to new climate schema (#24915)

* initial commit

* Update climate.py

pvizeli committed:
Migrate eq3btsmart

pvizeli committed:
Merge branch 'climate-1.0' of https://github.com/home-assistant/home-assistant into climate-1.0

b4dpxl commented:
Will the changes here address #20154? I've had a look around (not going to pretend I fully follow the code) but it doesn't appear as though it will. It looks as though it still calls async_set_temperature, which sets the target temperature even when away mode (or whatever it's called now) is enabled.

balloob committed:
Lint

pvizeli committed:
cleanup PRESET_MANUAL

Rendili commented:
@pvizeli , what is happening with Hive?

As far as I can tell, this PR is going to be a breaking change for Hive users, dropping hotwater support for Hive and change some standards in Hive climate compared to the rest of the Hive components which could break how they all work together, yet Hive is ticked at the top (as complete?).

kennedyshead commented:
I'll have a look at Melissa asap. What is the deadline for this?

balloob committed:
Migrate ecobee

balloob committed:
No conditional features

balloob commented:
@kennedyshead the plan is right now that we want to merge this as soon as all platforms have been migrated. It will require a bunch of testing in dev and beta.

marvin-w commented:
I had a look into the knx climate component: https://github.com/home-assistant/home-assistant/pull/24931

marvin-w committed:
KNX: Migrate climate component to new climate platform (#24931)

* Migrate climate component

* Remove unused code

* Corrected line length

* Lint

balloob committed:
Lint

ttroy50 commented:
Basic changes to the ephember component look ok. I'll try to get time to test it this evening.

Based on the hive component comments are hot water thermostats being moved from the climate component?

pvizeli committed:
fix tests

pvizeli commented:
Water heater aka boiler have his own component because they are no hvac like we want cover with climate component

pvizeli committed:
Fix value

zxdavb committed:
Migrate geniushub to new climate schema (#24191)

* Update one

* Fix model climate v2

* Cleanup p4

* Add comfort hold mode

* Fix old code

* Update homeassistant/components/climate/__init__.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

* Update homeassistant/components/climate/const.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

* First renaming

* Rename operation to hvac for paulus

* Rename hold mode to preset mode

* Cleanup & update comments

* Remove on/off

* Fix supported feature count

* Update services

* Update demo

* Fix tests & use current_hvac

* Update comment

* Fix tests & add typing

* Add more typing

* Update modes

* Fix tests

* Cleanup low/high with range

* Update homematic part 1

* Finish homematic

* Fix lint

* fix hm mapping

* Support simple devices

* convert lcn

* migrate oem

* Fix xs1

* update hive

* update mil

* Update toon

* migrate deconz

* cleanup

* update tesla

* Fix lint

* Fix vera

* Migrate zwave

* Migrate velbus

* Cleanup humity feature

* Cleanup

* Migrate wink

* migrate dyson

* Fix current hvac

* Renaming

* Fix lint

* Migrate tfiac

* migrate tado

* delinted

* delinted

* use latest client

* clean up mappings

* clean up mappings

* add duration to set_temperature

* add duration to set_temperature

* manual rebase

* tweak

* fix regression

* small fix

* fix rebase mixup

* address comments

* finish refactor

* fix regression

* tweak type hints

* delint

* manual rebase

zxdavb committed:
WIP: Fixes for honeywell migration to climate-1.0 (#24938)

* add type hints

* code tidy-up

ttroy50 commented:
I've checked the ephember component.

* Read / Set target temperature - OK
* Read current temperature - OK
* Read HVAC Mode - OK
* Set HVAC Mode -
  * No UI for this (In default lovelace card)
  * Works via services but seems to take 30 seconds to reflect in the UI
* Aux Mode
  * No UI for this
  * Works via services.
* Fan mode 
  * I see an empty Fan mode dropdown in the UI but Fan mode should not be supported by ephember

zxdavb commented:
@pvizeli If my system is in `HVAC_MODE_OFF` can `set_preset_mode()` be called to change the `PRESET` - which implies a change to the `hvac_mode()`) - or must the user first change to `HVAC_MODE_HEAT`?

fredrike commented:
I've tested the Daikin integration d0e457a and unfortunatly can't I set either fan-speed, fan-direction or ac_mode (heat/cool/dry etc.).

zxdavb committed:
Fixes for incomfort migration to climate-1.0 (#24936)

* delint type hints

* no async unless await

* revert: no async unless await

* revert: no async unless await 2

* delint

* fix typo

OnFreund commented:
I have the code needed to migrate coolmaster.
1. How do you add to this branch? No permissions
2. There's some serious loss of functionality - I can't turn on/off, or change mode (at least in the UI). Am I missing anything?

Jc2k committed:
Fix homekit_controller on climate-1.0 (#24948)

* Fix tests on climate-1.0 branch

* As part of climate-1.0, make state return the heating-cooling.current characteristic

* Fixes from review

pvizeli commented:
**The UI currently not updated or support the new schema. That follow next week**

OnFreund commented:
Thanks @pvizeli. I'm assuming the same also holds for google assistant and Alexa? Also, google assistant lets me turn my A/C on today - more functionality that will be lost after this PR is merged.

How do I add my coolmaster code here? A PR from my branch to this PR branch?

balloob commented:
Google Assistant and Alexa will support turn on if there are two hvac modes and one of those is off.

balloob committed:
lint

balloob committed:
Fix imports

pvizeli committed:
Migrate stibel_eltron

pvizeli committed:
Merge branch 'climate-1.0' of https://github.com/home-assistant/home-assistant into climate-1.0

pvizeli committed:
Fix lint

OnFreund commented:
I have a PR to merge into this branch here:
https://github.com/home-assistant/home-assistant/pull/24967
CI is failing both lint and tests, but neither seems to be related to my changes.

OnFreund committed:
Migrate coolmaster to climate 1.0 (#24967)

* Migrate coolmaster to climate 1.0

* fix lint errors

* More lint fixes

balloob committed:
Fix demo to work with UI

pvizeli committed:
Migrate spider

peternijssen commented:
@pvizeli I just was about to start working on Spider. Thanks for that :
Is the component now migrated? Should I still run a test?

balloob committed:
Demo update

balloob committed:
Updated frontend to 20190705.0

balloob committed:
Merge remote-tracking branch 'origin/dev' into climate-1.0

balloob commented:
@peternijssen yes still please run a test. The frontend has been updated and is part of this PR.

OnFreund commented:
@balloob I can confirm the UI now allows changing mode and turning off with Coolmaster. Still missing what I think is important functionality (on/off separate from mode), but I'll open a new arch issue for that once this PR is merged.

peternijssen commented:
> @peternijssen yes still please run a test. The frontend has been updated and is part of this PR.

All working as expected!

SukramJ committed:
Fix boost mode (#24980)

cgtobi committed:
Prepare Netatmo for climate 1.0 (#24973)

* Migration Netatmo

* Address comments

* Update climate.py

balloob committed:
Migrate ephember

Jc2k commented:
It looks like `homeassistant.components.homekit.type_thermostats` still needs updating. At least it still imports `ATTR_OPERATION_MODE` and `ATTR_OPERATION_LIST`. That integration doesnt seem to be on the checklist, should it be?

balloob committed:
Migrate Sensibo

Commit:
Implemented review comments (#24942)

balloob commented:
@Jc2k adding it

balloob committed:
Migrate ESPHome

balloob committed:
Migrate MQTT

balloob committed:
Migrate Nest

balloob committed:
Migrate melissa

frenck commented on homeassistant/components/radiotherm/climate.py:
```diff
@@ -306,34 +288,13 @@ def set_time(self):
+        elif hvac_mode == HVAC_MODE_HEAT:
             self.device.t_heat = self._target_temperature
-
-    def turn_away_mode_on(self):
```
Because this is removed, the whole away temp settings are no longer in a working state (which is fine I guess?)

However, all variables still exist, can be configured, however, are unused.

andrewsayre committed:
Initial/partial migration of ST

andrewsayre committed:
Merge branch 'climate-1.0' of https://github.com/home-assistant/home-assistant into climate-1.0

andrewsayre committed:
Migrate ST

andrewsayre commented:
Few issues/comments I see with the new schema:
1. Missing `CURRENT_HVAC_FAN_ONLY` when we have `HVAC_MODE_FAN_ONLY`.
1. Missing preset equivalents for `turn_on` and `turn_off` when the device has a power switch separate from HVAC mode
1. Given we are using `auto` to mean a preset and `heat_cool` to mean what thermostats call `auto`, I think there's going to be a lot of confusion...  The updated front-end is showing `heat_cool` with the `auto` label.  Setting HVAC mode _to_ `auto` seems problematic too, given it could mean any number of presets, maybe it shouldn't be allowed?

SukramJ committed:
Remove Away mode (#24995)

pvizeli commented:
@andrewsayre `auto` means like in the `const.py` description, the device is controlled by not us and we can't modification the temperature because that is done by hub/AI/whatever.

The focus of climate has changed now that we have our own logic. So you need now adapt our logic of HVAC to your interface instead the HA is adapt the logic from your device. That means, there is no turn on/off because you need to turn your device on the interface if they change the mode or expose it also as a switch. So if we move to HVAC_OFF, you turn the device off and if they move to heat or whatever you switch your device also on.

zxdavb committed:
Migrate evohome, cache access tokens (#24491)

* add water_heater, add storage - initial commit

* add water_heater, add storage - initial commit

delint

add missing code

desiderata

update honeywell client library & CODEOWNER

add auth_tokens code, refactor & delint

refactor for broker

delint

* Add Broker, Water Heater & Refactor

add missing code

desiderata

* update honeywell client library & CODEOWNER

add auth_tokens code, refactor & delint

refactor for broker

* bugfix - loc_idx may not be 0

more refactor - ensure pure async

more refactoring

appears all r/o attributes are working

tweak precsion, DHW & delint

remove unused code

remove unused code 2

remove unused code, refactor _save_auth_tokens()

* support RoundThermostat

bugfix opmode, switch to util.dt, add until=1h

revert breaking change

* store at_expires as naive UTC

remove debug code

delint

tidy up exception handling

delint

add water_heater, add storage - initial commit

delint

add missing code

desiderata

update honeywell client library & CODEOWNER

add auth_tokens code, refactor & delint

refactor for broker

add water_heater, add storage - initial commit

delint

add missing code

desiderata

update honeywell client library & CODEOWNER

add auth_tokens code, refactor & delint

refactor for broker

delint

bugfix - loc_idx may not be 0

more refactor - ensure pure async

more refactoring

appears all r/o attributes are working

tweak precsion, DHW & delint

remove unused code

remove unused code 2

remove unused code, refactor _save_auth_tokens()

support RoundThermostat

bugfix opmode, switch to util.dt, add until=1h

revert breaking change

store at_expires as naive UTC

remove debug code

delint

tidy up exception handling

delint

* update CODEOWNERS

* fix regression

* fix requirements

* migrate to climate-1.0

* tweaking

* de-lint

* TCS working? & delint

* tweaking

* TCS code finalised

* remove available() logic

* refactor _switchpoints()

* tidy up switchpoint code

* tweak

* teaking device_state_attributes

* some refactoring

* move PRESET_CUSTOM back to evohome

* move CONF_ACCESS_TOKEN_EXPIRES CONF_REFRESH_TOKEN back to evohome

* refactor SP code and dt conversion

* delinted

* delinted

* remove water_heater

* fix regression

Rendili commented:
@pvizeli Maybe I am not understanding your last message, but, if set to auto on Hive, you can still change the temperature and it will set to that temperature (as designed and as with the official Hive app), when the next schedule arrives in auto the Hive system will automatically change to that scheduled temperature.
So in summary for Hive at least, we should still be allowed to change the temperature in Auto mode.

pvizeli committed:
Migrate homekit

zxdavb commented:
@Rendili 
> So in summary for Hive at least, we should still be allowed to change the temperature in Auto mode.

I think I have this right:
 - `HVAC_MODE_HEAT` means will 'automatically' heat or not (say to a schedule, or an occupancy sensor), but the user can change the temperature
 - `HVAC_MODE_AUTO` means will 'automatically' heat/cool or not (say to a schedule, or an occupancy sensor), but the user cannot override the target temperature (although they may be able to change the `HVAC_MODE_*`

I have an Honeywell evohome system that is multi-zone (multiple rooms, each with a TRV that can be controlled individually) with a central hub/controller (house).  The hub can be set to various modes, e.g. **Auto**, **Eco**, **Away** & **Off**.

In **Auto** mode, each zone (room) follows its own schedule, unless the user changes the setpoint in that room. In **Away** mode, the zone will have it's target temperature fixed at 12C & the user cannot override that temp (they're not at home, in any case).

When the hub is in:
 - **Auto** mode, each zone will be `HVAC_MODE_HEAT` (with no preset)
 - **Away** mode, each zone will be `HVAC_MODE_AUTO` (with `PRESET_AWAY`)

Rendili commented:
@zxdavb 
Thanks for the expanding more on this.

Hive has a Heat mode, which has no schedule and just one set temperature, do you know how this would fit in to the new system?

Hive has:
Heat - The zone will maintain a set constant temperature and no schedule
Auto - The zone will maintain the temperature as set out in a schedule
Off - The zone will not heat at all

zxdavb commented:
> @zxdavb
> Thanks for the expanding more on this.
> 
> Hive has a Heat mode, which has no schedule and just one set temperature, do you know how this would fit in to the new system?
> 
> Hive has:
> Heat - The zone will maintain a set constant temperature and no schedule
> Auto - The zone will maintain the temperature as set out in a schedule
> Off - The zone will not heat at all

I assume you have a UK-based Hive that turns the boiler on/off & the rads have their own dumb TRVs?  We are not talking about the multi-zone feature of Hive?
 - **Heat** - `HVAC_MODE_HEAT` and `PRESET_MANUAL`
 - **Auto** - `HVAC_MODE_HEAT` and no preset (None)
 - **Off** - `HVAC_MODE_OFF`

Note that `PRESET_MANUAL` is not yet in **const.py**, but that wouldn't stop you from adding one to the integration itself.

Jc2k commented:
Ah, looks like homekit_controller is still wrong too, then. It has 4 modes you can set, from the spec:

 * **Off**
 * **Heat**: If the current temperature is below the target temperature then turn on heating.
 * **Cool**: If the current temperature is above the target temperature then turn on cooling.
 * **Auto**: Turn on heating or cooling to maintain temperature within the heating and cooling threshold of the target temperature.

You can **always** set the target temperature.

Not really sure how that maps here.

pvizeli committed:
Cleanup away mode

Rendili commented:
> @zxdavb
> I assume you have a UK-based Hive that turns the boiler on/off & the rads have their own dumb TRVs?  We are not talking about the multi-zone feature of Hive?
> 
> Heat - HVAC_MODE_HEAT and PRESET_MANUAL
> Auto - HVAC_MODE_HEAT and no preset (None)
> Off - HVAC_MODE_OFF
> 
> Note that PRESET_MANUAL is not yet in const.py, but that wouldn't stop you from adding one to the integration itself.

Yes exactly, the Hive component was written initially for the UK Hive system (but may work with the other countries versions?).
It currently supports a single zone installation and multi-zone installations.
With a single zone installation you currently get a single climate device in HA and if your Hive installation supports HotWater (like mine), you also get a climate device acting as the hotwater device (this was the closest thing in HA at the time of writing the component).
With a multi zone installation I believe you get an HA climate device per zone plus any hotwater support.

At the moment the new Hive TRV are not supported, so the only dump TRVs are assumed today.

In HA each climate device represents a zone and each have the three modes; `Heat`, `Auto` & `Off` and each can have their temperature set in `Heat` & `Auto` mode.
Also each can have the boost invoked in `Heat` & `Auto` mode.

Sounds like i need to have a play with the latest code and see what is going on and how it maps to the real life Hive system. I am hoping i will get some time tonight, but not sure i will have enough time to rebuild my dev / test sytem and test.

pvizeli committed:
Fix tests

pvizeli committed:
add helpers

zxdavb commented:
@Jc2k 
> Ah, looks like homekit_controller is still wrong too, then. It has 4 modes you can set, from the spec:
> 
> * **Off**
> * **Heat**: If the current temperature is below the target temperature then turn on heating.
> * **Cool**: If the current temperature is above the target temperature then turn on cooling.
> * **Auto**: Turn on heating or cooling to maintain temperature within the heating and cooling threshold of the target temperature.
> 
> You can **always** set the target temperature.

I believe it is:
* **Off**: `HVAC_MODE_OFF`
* **Heat**: `HVAC_MODE_HEAT`
* **Cool**: `HVAC_MODE_COOL`
* **Auto**: `HVAC_MODE_HEAT_COOL`

pvizeli committed:
fix tests melissa

pvizeli committed:
Fix nehueat

pvizeli committed:
fix zwave

pvizeli committed:
add more tests

pvizeli committed:
fix deconz

pvizeli committed:
Fix climate test emulate_hue

pvizeli committed:
fix tests

pvizeli committed:
fix dyson tests

pvizeli committed:
fix demo with new layout

pvizeli committed:
fix honeywell

Jc2k commented:
@zxdavb cheers!

Have submitted #25009 to fix that.

Jc2k committed:
Switch homekit_controller to use HVAC_MODE_HEAT_COOL instead of HVAC_MODE_AUTO (#25009)

balloob committed:
Lint

balloob committed:
PyLint

balloob committed:
Pylint

pvizeli committed:
fix fritzbox tests

pvizeli committed:
Fix google

pvizeli committed:
Fix all tests

pvizeli committed:
Fix lint

pvizeli committed:
Fix auto for homekit like controler

pvizeli committed:
Fix lint

pvizeli committed:
fix lint

balloob commented:
🎉 🎉 🎉 🎉 🎉 

We did it! We made a lots of changes and things might have been broken. If you find a bug in one of the integrations or have a PR with improvements, please tag PRs with `[climate]` so it is easy to spot for us. Feel free to tag me and I prioritize reviewing.

Rendili commented:
Hi @balloob,

Khole and I are working on fixes to the Hive climate code, which we think we have completed now and re-adding Hive hotwater support as water_heater after it was removed from climate.

We are still ironing out some bugs in the water_heater, but if not complete before Climate 1.0 goes live we will lose hotwater / water_heater support for Hive (assuming no one else is working on this)

How do you recommend we proceed? We were aiming for one PR with the new file for water_heater and the fixed file for climate.
and do we now make the PR for DEV rather than Climate branch?

elupus commented:
The ecobee failure is due to dict in 3.5 not preserving order on keys.
https://github.com/home-assistant/home-assistant/blob/ab832cda710f9bf14111b438005ed27aca7cad0d/homeassistant/components/ecobee/climate.py#L427

balloob commented:
@Rendili PRs against dev branch. We are cutting the 96 beta this Wednesday but will include a water heater PR in 96 if it comes in during the beta period.

@elupus good catch, I'll open a PR with a fix.

balloob commented:
@elupus https://github.com/home-assistant/home-assistant/pull/25019
